R/dataframe-frame.R

```diff
@@ -2200,3 +2200,114 @@ dataframe__group_by_dynamic <- function(
     )
   })
 }
+
+#' Unstack a long table to a wide form without doing an aggregation
+#'
+#' This can be much faster than a pivot, because it can skip the grouping phase.
+#'
+#' @param ... <[`dynamic-dots`][rlang::dyn-dots]> Column name(s) and selector(s)
+#' to include in the operation. If `NULL` (default), use all columns.
+#' @param step Number of rows in the unstacked frame.
+#' @param how Direction of the unstack. Must be one of `"vertical"` or
+#' `"horizontal"`.
+#' @param fill_values Fill values that don't fit the new size with this value.
+#' This can be a scalar value, an Expr, or a named list of the sort
+#' `list(<column_name> = <fill_value>)`. See examples.
+#'
+#' @inherit as_polars_df return
+#' @examples
+#' df <- pl$DataFrame(x = LETTERS[1:8], y = 1:8)$with_columns(
+#'   z = pl$int_ranges(pl$col("y"), pl$col("y") + 2, dtype = pl$UInt8)
+#' )
+#' df
+#'
+#' df$unstack(step = 4, how = "vertical")
+#' df$unstack(step = 2, how = "horizontal")
+#' df$unstack(cs$numeric(), step = 5, fill_values = 0)
+#' df$unstack("x", "y", step = 5, fill_values = list(y = 999, x = "foo"))
+dataframe__unstack <- function(
+  ...,
+  step,
+  how = c("vertical", "horizontal"),
+  fill_values = NULL
+) {
+  wrap({
+    check_dots_unnamed()
+
+    is_named_list <- is.list(fill_values) &&
+      !is.null(names(fill_values)) &&
+      any(names(fill_values) != "")
+    is_scalar_or_expr <- length(fill_values) == 1 || is_polars_expr(fill_values)
+    if (!is.null(fill_values) && !is_named_list && !is_scalar_or_expr) {
+      abort("`fill_value` must be a scalar or a named list.")
+    }
```

Comment:
I don't think we have a helper to check if input is a named list or a scalar (and I don't think it's worth doing one)